File: Wave Equation/CircularMembrane.m
function CircularMembrane
clc
a=0.5;
r=3;
rho=0:r/50:r;
phi=0:pi/25:2*pi;

tmax=30;
t=linspace(0,tmax);

N=40;

% namirane na purvite 40 polojitelni nuli na funkciata na Bessel

bessj0 = inline('besselj(0,x)') ;
for n = 1:N
mju(n) = fzero(bessj0, [( n-1) n ]* pi );
end

% zadavane na nachalnoto polojenie

function y=tau(rho)
            y=rho.^2.*sin(pi*rho).^3;
end

% presmiatane stoinostite na reshenieto
    
function y=solution(R,t)
y=0;
    for m=1:N
    s=tau(R(1,:)).*R(1,:).*besselj(0,mju(m)*R(1,:)/r);
    A0m=4*trapz(R(1,:),s)/((r^2)*(besselj(1,mju(m))^2)); 
    y=y+A0m*cos(a*mju(m)*t/r)*besselj(0,mju(m)*R/r);
    end
end

% suzdavane na mreja ot tocki

[R,p]=meshgrid(rho,phi);
X=R.*cos(p);Y=R.*sin(p);

% animacia na dvijenieto na membranata
 
for k=1:length(t)
    Z=solution(R,t(k));
surf(X,Y,Z)
axis([-r r -r r -30 30])
title('Circular membrane')
xlabel('x');
ylabel('y');
zlabel('u(x,y,t)');
M(k)=getframe;
end

   %movie(M,1,5)    
end

File: Wave Equation/circular_membrane.py
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.animation import FuncAnimation, PillowWriter
from mpl_toolkits.mplot3d import Axes3D
from scipy.optimize import root_scalar
from scipy.special import jv as besselj


def circular_membrane():
    a = 0.5
    r = 3
    rho = np.linspace(0, r, 51)
    phi = np.linspace(0, 2 * np.pi, 51)

    tmax = 30
    t = np.linspace(0, tmax, 100)
    N = 40

    # Find the first 40 positive zeros of the Bessel function J0
    mju = []
    for n in range(1, N + 1):
        zero = root_scalar(
            lambda x: besselj(0, x), bracket=[(n - 1) * np.pi, n * np.pi]
        )
        mju.append(zero.root)
    mju = np.array(mju)

    # Define the initial position function
    def tau(rho):
        return rho**2 * np.sin(np.pi * rho) ** 3

    # Solution function
    def solution(R, t):
        y = np.zeros_like(R)
        for m in range(N):
            s = tau(R[0, :]) * R[0, :] * besselj(0, mju[m] * R[0, :] / r)
            A0m = 4 * np.trapz(s, R[0, :]) / ((r**2) * (besselj(1, mju[m]) ** 2))
            y += A0m * np.cos(a * mju[m] * t / r) * besselj(0, mju[m] * R / r)
        return y

    # Create a grid of points
    R, p = np.meshgrid(rho, phi)
    X = R * np.cos(p)
    Y = R * np.sin(p)

    # Set up the figure and axis for animation
    fig = plt.figure()
    ax = fig.add_subplot(111, projection="3d")
    ax.set_xlim(-r, r)
    ax.set_ylim(-r, r)
    ax.set_zlim(-30, 30)
    ax.set_title("Circular membrane")
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_zlabel("u(x,y,t)")

    # Update function for FuncAnimation
    def update(frame):
        ax.clear()
        Z = solution(R, frame)
        # ax.plot_surface(X, Y, Z, cmap="viridis")
        ax.plot_surface(X, Y, Z, cmap="viridis", vmin=-30, vmax=30)
        ax.set_xlim(-r, r)
        ax.set_ylim(-r, r)
        ax.set_zlim(-30, 30)
        ax.set_title("Circular membrane")
        ax.set_xlabel("x")
        ax.set_ylabel("y")
        ax.set_zlabel("u(x,y,t)")

    # Create the animation
    anim = FuncAnimation(fig, update, frames=t, interval=50)

    # Save the animation
    # anim.save("circular_membrane_animation.mp4", writer="ffmpeg", fps=20)
    # Alternatively, for GIF format:
    anim.save("circular_membrane_animation.gif", writer="imagemagick", fps=20)

    plt.show()


# Call the function to execute the animation
circular_membrane()
